### DISCUSSION TITLE
Localization with properties files

### DISCUSSION THREAD

**Discussion by gregjotau** [jstachio/jstachio#343] [General]
For thymeleaf I simply use: th:text="#{exchange}

and then exchange property defined in a properties file, will something similar be implemented in: 

https://github.com/jstachio/jstachio/issues/104

That is one of the reason I stick with thymeleaf for now.

**Comment by agentgt** [comment]
Do you want it exactly how Thymeleaf works or are there somethings you would change?

I will add this to 1.4.0 . Thanks for the comment to motivate me to finally do something about 😄

**Reply by gregjotau** [reply]
Since I already have translations in property files and it works well I would say yes. But if there is a better way to do things that I am not aware of / more people use, focus on that instead :) 

But all java projects I have been working with have the translations in property files, also intellisense hopefully would work the same as now where non-existing keys leads to error:

<img width="708" alt="image" src="https://github.com/jstachio/jstachio/assets/13047497/d474d133-bebb-45cd-bece-cccb5dc03c52">

And CTRL+CLICK into the properties file to go to translations: 

<img width="514" alt="Screenshot 2024-04-01 at 09 25 57" src="https://github.com/jstachio/jstachio/assets/13047497/073a6e3d-69ff-4f34-be01-cf84dbb9c376">

Not sure if that would work out of the box with mustache files if there is a standard for this in Mustache language that the mustache plugin support.

**Reply by agentgt** [reply]
The trickiest part to i18n for JStachio is parameterization of the i18n message.

I do have an example doing mustache.java style: https://github.com/jstachio/jstachio/tree/main/test/examples/src/main/java/io/jstach/examples/i18n

But what I'm looking into is compile time checking of the property keys similar to this library:
https://santhosh-tekuri.github.io/jlibs/i18n/Internationalization.html

However that still has the problem of parameterization and how Mustache lacks a way to call something with more than one parameter.

**Reply by dsyer** [reply]
What about named parameters, and ignoring the names when the output is processed? So user can write template like it was a normal fragment, and just enumerate parameters with arbitrary names.

**Reply by agentgt** [reply]
I think I know what you are saying it is genius but can you elaborate some more?

I really like the idea of treating each entry in the property file as a partial or even better parent partial.

Then for parameters you convert the {0} to {{$0}}{{/0}}.

Edit I guess I can’t use blocks for parameters because we need the actual object like number and date for formatting and can’t just be string based.

**Reply by dsyer** [reply]
The terminology is hard for me to keep straight. Yes, I think “parent partial” was what I had in mind. I leave the syntax up to you, but some examples would probably help firm up the idea.

BTW I liked the way `{{@context}}` got added a while back, and I was wondering then if there might be a `{{@message}}` (or `{{>@message}}` maybe) at some point to address this gap.

**Reply by dsyer** [reply]
I suppose it might end up looking like this (borrowing an example from the Thymeleaf docs):

```
<p>
  {{#@message}}{{<home.welcome}}{{$0}}{{session.user.name}}{{/0}}{{/home.welcome}}{{/@message}}
</p>
```

Slightly clunky. Maybe `{{<@message.home.welcome}}` might work?

**Reply by agentgt** [reply]
Yes I think we are on the same page and I can see it working for String-like localization but the question is how to deal with dates and currency etc.


```hbs
  {{#@message}}{{<amountOwed}}{{$0}}{{balance}}{{/0}}{{/amountOwed}}{{/@message}}
```

message_en.properties:

```properties
amountOwed=You owe {0,number,#.##}.
```

Basically:

```
{{$0}}{{balance}}{{/0}}
```

As you know is not like an object parameter but more of a template. I know its inside a lambda but it seems disingenuous to make it do objects instead of template replacement. 

Since `{{#@message}}` is inherently a magic lambda and lambdas can have anything I'm inclined ATM to make it a streamlined syntax where its clear localization/i18n is happening over the confusion of parent template usage.

If it weren't for the overwhelming infrastructure and standard practice of resource bundles using `MessageFormat` syntax I would jump all over the partial technique and that is why I'm leaning towards: 

```hbs
  {{#@message}}home.welcome(session.user.name){{/@message}}
```

Or possibly this:

```hbs
  {{! First dot is required and params CSV}}
  {{#@message.home.welcome}}session.user.name,session.time{{/@message.home.welcome}}
```

One little problem is the off chance that someone wants a i18n message to not be escaped perhaps because its a javascript portion of the template (lets ignore how that is probably a bad idea).

The obvious would be to triple mustache the lambda but the spec isn't clear on if that is allowed.

The easy solution I suppose is to add two magic/static message lambdas one that will escape and one that will not: eg. `@message_raw` or something (bike shed name later).

**Reply by agentgt** [reply]
Another issue is JMustache compatibility. Whatever we choose I want to make sure its possible to add similar capability to JMustache ideally as an extension w/o changing JMustache core.

While JMustache allows stack resolution for objects by index it does not have name based lookup. So that would have to be added to Mustache.Lambda.

**Reply by agentgt** [reply]
https://github.com/jstachio/jmustache/tree/fragment_name_lookup
